### PR TITLE
Add command-line control over workspace actions

### PR DESCRIPTION
**Pull request by douglasjacobsen**
This merge adds the ability for workspace actions to filter which experiments they should act on within the workspace. This is performed through the `--where` and `--exclude-where` arguments.

Additionally, the logic around pipelines is overhauled in this merge to make it easier to extend and manipulate the pipelines.

### CODE REVIEW

**Comment by douglasjacobsen** [comment]
I need to update the docs for this, but I think the code is ready to review now.

**Comment by douglasjacobsen** [comment]
It would probably be good to get an extra thorough review of https://github.com/GoogleCloudPlatform/ramble/pull/258/files#diff-30efb4962f087cdf7750874e43ce62c7a1475c75cf7f1f2d315ccf6dbc22e398R586-R607

And it might be good to have a discussion around if we want hashes to be computed on dry-runs or not.

Currently, the logic as implemented will allow the hash to change between a dry-run and a full run of `ramble workspace setup`.

File: lib/ramble/ramble/pipeline.py [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+        self.upload_results = upload
+
+    def _prepare(self):
+        super()._construct_hash()
+        super()._prepare()
+
+    def _complete(self):
+        self.workspace.dump_results(output_formats=self.output_formats)
+
+        # FIXME: this will fire the analyze logic of twice currently
```
Comment (rfbgo) [inline lib/ramble/ramble/pipeline.py:186]: This is in the wrong place, I think (or at least the upload_results doesn't trigger analyze) 

We can probably just remove it..

**Review by rfbgo — CHANGES_REQUESTED** [review]
Awesome contribution! Few small areas of feedback

File: lib/ramble/ramble/cmd/workspace.py [hunk 2]
```diff
@@ -317,18 +319,28 @@ def workspace_setup_setup_parser(subparser):
...
     if args.dry_run:
         ws.dry_run = True
 
+    filters = ramble.filters.Filters(
+        phase_filters=args.phases,
+        include_where_filters=args.where,
+        exclude_where_filters=args.exclude_where
+    )
+
+    pipeline_cls = ramble.pipeline.pipeline_class(current_pipeline)
```
[the remark this answers is not in the document]
Comment (douglasjacobsen) [reply lib/ramble/ramble/cmd/workspace.py:338]: This actually does return a class, but the comment in the factory was wrong. I updated it in the latest commit.

**Review by rfbgo — APPROVED** [review]

**Pull request merged**